tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp

```diff
@@ -54,6 +56,120 @@ extract_dr_bondi_scalars_from_cartesian_metric(
     const CartesianiSphericalJ& inverse_jacobian,
     const CartesianiSphericalJ& dr_inverse_jacobian,
     double extraction_radius) noexcept;
+
+// This function checks the consistency of the 3+1 ADM quantities in the
+// `boundary_tuple` with quantities computed from `expected_spacetime_metric`,
+// `expected_dt_spacetime_metric`, and `expected_d_spacetime_metric`. If the
+// expected quantities are also extracted from the tuple, this simply checks
+// that the boundary computation has produced a consistent set of quantities
+// and has not generated NaNs or other pathological values in the process.
+template <typename... TupleTags>
+void check_adm_metric_quantities(
+    const tuples::TaggedTuple<TupleTags...>& boundary_tuple,
+    const tnsr::aa<DataVector, 3>& expected_spacetime_metric,
+    const tnsr::aa<DataVector, 3>& expected_dt_spacetime_metric,
+    const tnsr::iaa<DataVector, 3>& expected_d_spacetime_metric) noexcept {
+  // check the 3+1 quantities are computed correctly in the abstract base class
+  // `WorldtubeData`
+  const auto& dr_cartesian_coordinates =
+      get<Tags::Dr<Tags::CauchyCartesianCoords>>(boundary_tuple);
+
+  const auto& lapse = get<gr::Tags::Lapse<DataVector>>(boundary_tuple);
+  const auto& shift =
+      get<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>(boundary_tuple);
+  const auto& spatial_metric =
+      get<gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>(
+          boundary_tuple);
+
+  const auto expected_spatial_metric =
+      gr::spatial_metric(expected_spacetime_metric);
+  const auto expected_inverse_spatial_metric =
+      determinant_and_inverse(expected_spatial_metric).second;
+  const auto expected_shift =
+      gr::shift(expected_spacetime_metric, expected_inverse_spatial_metric);
+  const auto expected_lapse =
+      gr::lapse(expected_shift, expected_spacetime_metric);
+  CHECK_ITERABLE_APPROX(spatial_metric, expected_spatial_metric);
+  CHECK_ITERABLE_APPROX(shift, expected_shift);
+  CHECK_ITERABLE_APPROX(lapse, expected_lapse);
+
+  const auto& pi =
+      get<GeneralizedHarmonic::Tags::Pi<3, ::Frame::Inertial>>(boundary_tuple);
+  const auto dt_spacetime_metric_from_pi =
+      GeneralizedHarmonic::time_derivative_of_spacetime_metric(
+          expected_lapse, expected_shift, pi, expected_d_spacetime_metric);
+  CHECK_ITERABLE_APPROX(expected_dt_spacetime_metric,
+                        dt_spacetime_metric_from_pi);
+
+  const auto& dt_lapse =
```

Comment:
why do we need this block, how can it not be true if the pi is correct?

Reply:
These quantities are also retrieved from the `boundary_tuple`, so while it seems pretty redundant for the current implementation, it's not necessarily clear that the boundary calculation needs to compute these dt values from the `pi`, so it seemed best to include them in the general utility.

Reply:
Sounds good! Maybe add a comment just so that they don't get deleted in the future?